homeassistant/components/bkk/sensor.py

```diff
@@ -0,0 +1,311 @@
+        self._departure_time = (
+            current_stop_time["departureTime"]
+            if "departureTime" in current_stop_time
+            else 0
+        )
+        self._predicted_departure_time = (
+            current_stop_time["predictedDepartureTime"]
+            if "predictedDepartureTime" in current_stop_time
+            else 0
+        )
+        self._line = routes[trips[current_stop_time["tripId"]]["routeId"]]["shortName"]
+        self._id = routes[trips[current_stop_time["tripId"]]["routeId"]]["id"]
+        self._name = (
+            f"{self.line}->{trips[current_stop_time['tripId']]['tripHeadsign']}"
+        )
+        self._attributes["line"] = self._line
+        self._attributes["route_id"] = self._id
+        self._attributes["wheelchair_accessible"] = trips[current_stop_time["tripId"]][
+            "wheelchairAccessible"
+        ]
+        self._attributes["destination"] = trips[current_stop_time["tripId"]][
+            "tripHeadsign"
+        ]
+        self.calculate_state()
+
+    @property
+    def icon(self):
+        """Return entity icon."""
+        return self._icon
+
+    @property
+    def state(self):
+        """Return route state."""
+        return self._state
+
+    @property
+    def device_state_attributes(self):
+        """Return state attributes."""
+        return self._attributes
+
+    @property
+    def name(self):
+        """Return entity name."""
+        return self._name
+
+    @property
+    def unit_of_measurement(self):
+        """Return measurement unit."""
+        return "min"
+
+    @property
+    def unique_id(self):
+        """Return unique id."""
+        return f"{self._stop_id}_{self._entity_idx}"
+
+    def update(self):
+        """Update entity state."""
+        pass
+
+    @staticmethod
+    def to_iso_datetime(epoch):
+        """Convert epoch date format to ISO format."""
+        if epoch:
+            return datetime.isoformat(datetime.fromtimestamp(epoch))
+        return None
+
+    @staticmethod
+    def calc_state_value(now, p_time):
+        """Return time between now and a timestamp in minutes."""
+        return int((datetime.fromtimestamp(p_time) - now).total_seconds() / 60)
+
+    def calculate_state(self):
+        """Calculate entity state, return value in minutes."""
+        now = datetime.now()
+        state = None
+        t_next = datetime.fromtimestamp(self._departure_time)
+        t_predicted = datetime.fromtimestamp(self._predicted_departure_time)
+        if t_next > now < t_predicted:
+            predicted = self.calc_state_value(now, self._predicted_departure_time)
+            next_time = self.calc_state_value(now, self._departure_time)
+            state = min(next_time, predicted)
+        elif t_next < now < t_predicted:
+            state = self.calc_state_value(now, self._predicted_departure_time)
+        elif t_next > now > t_predicted:
+            state = self.calc_state_value(now, self._departure_time)
```

Comment:
These entities should use the `"timestamp"` `device_class` and return an absolute utc isoformatted timestamp as state.

https://developers.home-assistant.io/docs/entity_sensor#available-device-classes